pd/src/state.rs

```diff
@@ -232,6 +233,58 @@ INSERT INTO notes (
         })
     }
 
+    /// Retreive the current validator set.
+    ///
+    pub async fn validators(&self) -> Result<BTreeMap<tendermint::PublicKey, Validator>> {
+        let mut conn = self.pool.acquire().await?;
+
+        let mut validators: BTreeMap<tendermint::PublicKey, Validator> = BTreeMap::new();
+
+        let stored_validators = query!(r#"SELECT tm_pubkey, voting_power FROM validators"#)
+            .fetch_all(&mut conn)
+            .await?;
+        for row in stored_validators.iter() {
+            // NOTE: we store the validator's public key in the database as a json-encoded string,
+            // because Tendermint pubkeys can be either ed25519 or secp256k1, and we want a
+            // non-ambiguous encoding for the public key.
+            let decoded_pubkey: tendermint::PublicKey = serde_json::from_slice(&row.tm_pubkey)?;
+
+            // NOTE: voting_power is stored in the psql database as a `bigint`, which maps to an
+            // `i64` in sqlx. try_into uses the `TryFrom<i64>` implementation for voting power from
+            // Tendermint, so will return an error if voting power is negative (and not silently
+            // overflow).
+            validators.insert(
+                decoded_pubkey,
+                Validator::new(decoded_pubkey, row.voting_power.try_into()?),
+            );
+        }
+
+        Ok(validators)
+    }
+
+    /// set the initial validator set, inserting each validator in `validators` into the state.
+    pub async fn set_initial_validators(
+        &self,
+        validators: &BTreeMap<tendermint::PublicKey, Validator>,
+    ) -> Result<()> {
```

Comment:
non-blocking on this PR: we have some centralization of our database writes into the `commit` method, maybe it would make sense to merge this setup code with the setup code in #234 into a single method that would be more clearly a call-once method?